File: src/main/java/dark/leech/text/plugin/api/RepositorySource.java
package dark.leech.text.plugin.api;

import java.util.concurrent.CompletableFuture;

import dark.leech.text.enities.PluginEntity;

/**
 * Abstraction for plugin repository sources. Supports multiple repository types (vBook, LeechText,
 * custom).
 *
 * @param <T> Metadata type for this repository
 */
public interface RepositorySource<T extends PluginMetadata> {

    /**
     * Unique identifier for this repository source. Format: "{type}-{id}" e.g., "vbook-default",
     * "leechtext-community"
     */
    String getId();

    /** Human-readable name for display in UI. */
    String getName();

    /** Repository type identifier. Examples: "vbook", "leechtext", "github", "custom" */
    String getType();

    /**
     * Check if this repository source is enabled. Disabled sources are not searched or downloaded
     * from.
     */
    boolean isEnabled();

    /** Enable or disable this repository source. */
    void setEnabled(boolean enabled);

    /**
     * Search for plugins matching the given query. Results are paginated to handle large
     * repositories.
     *
     * @param query Search query string
     * @param page Page number (0-indexed)
     * @param pageSize Number of results per page (recommended: 20-50)
     * @return Paginated search results
     * @throws RepositoryException if search fails
     */
    PaginatedResult<T> search(String query, int page, int pageSize) throws RepositoryException;

    /**
     * List all available plugins in this repository.
     *
     * @param page Page number (0-indexed)
     * @param pageSize Results per page
     * @return Paginated list of all plugins
     * @throws RepositoryException if listing fails
     */
    PaginatedResult<T> listAll(int page, int pageSize) throws RepositoryException;

    /**
     * Download a plugin by its identifier. Returns raw plugin entity that needs format conversion.
     *
     * @param pluginId Unique plugin identifier within this repository
     * @return Downloaded plugin entity
     * @throws RepositoryException if download fails
     */
    PluginEntity download(String pluginId) throws RepositoryException;

    /**
     * Check if this source can handle the given URL. Used for auto-discovery functionality.
     *
     * @param url URL to check
     * @return true if this source can provide a plugin for the URL
     */
    boolean canHandle(String url);

    /**
     * Find and install plugin for the given URL. Asynchronous operation - returns immediately with
     * Future.
     *
     * @param url URL to match against plugins
     * @return Future that completes with installed plugin, or null if no match
     */
    CompletableFuture<PluginEntity> findAndInstall(String url);

    /**
     * Get health status of this repository. Used for monitoring and circuit breaking.
     *
     * @return Health status
     */
    RepositoryHealth getHealth();

    /** Get repository metadata. Includes version, update check URL, etc. */
    RepositoryMetadata getMetadata();
}

File: engine/src/sharedJvmMain/kotlin/dev/haipham22/leechtext/plugin/api/RepositorySource.kt
package dev.haipham22.leechtext.plugin.api

import dev.haipham22.leechtext.entities.PluginEntity
import java.io.InputStream
import java.util.concurrent.CompletableFuture

/**
 * Abstraction cho plugin repository sources (port từ api/RepositorySource.java) — hỗ trợ
 * nhiều loại repo (vBook, LeechText, custom).
 */
interface RepositorySource<T : PluginMetadata> {
    /** Unique id — format "{type}-{id}" vd "vbook-default". */
    val id: String

    /** Tên hiển thị trong UI. */
    val name: String

    /** Loại repository — vd "vbook", "leechtext", "github", "custom". */
    val type: String

    /** Source có đang enabled không — disabled không search/download. */
    var isEnabled: Boolean

    /**
     * Search plugin theo query, kết quả phân trang.
     *
     * @param query query string
     * @param page page number (0-indexed)
     * @param pageSize results per page (khuyến nghị 20-50)
     */
    @Throws(RepositoryException::class)
    fun search(
        query: String?,
        page: Int,
        pageSize: Int,
    ): PaginatedResult<T>

    /** List mọi plugin trong repository (phân trang). */
    @Throws(RepositoryException::class)
    fun listAll(
        page: Int,
        pageSize: Int,
    ): PaginatedResult<T>

    /** Download plugin theo id — trả raw PluginEntity cần format conversion. */
    @Throws(RepositoryException::class)
    fun download(pluginId: String): PluginEntity

    /** Source có handle được URL này không — dùng cho auto-discovery. */
    fun canHandle(url: String?): Boolean

    /** Tìm + install plugin cho URL — async, trả null nếu không match. */
    fun findAndInstall(url: String?): CompletableFuture<PluginEntity?>

    /** Health status — dùng cho monitoring/circuit breaking. */
    val health: RepositoryHealth

    /** Metadata repository — version, update check URL, etc. */
    val metadata: RepositoryMetadata
}

/**
 * Abstraction cho plugin file format (port từ api/PluginFormat.java) — mỗi repository có thể
 * dùng format khác nhau.
 */
interface PluginFormat {
    /** Format id — vd "vbook-json", "leechtext-json", "zip". */
    val formatType: String

    /**
     * Parse plugin từ input stream, validate format khi parse.
     */
    @Throws(PluginFormatException::class)
    fun parse(input: InputStream): PluginEntity

    /** Validate plugin trước khi install — required fields, script syntax, security. */
    fun validate(plugin: PluginEntity): ValidationResult

    /** Convert sang standard LeechText format nếu cần (field mapping). */
    fun normalize(plugin: PluginEntity): PluginEntity

    /**
     * Extract metadata không full parse — dùng cho search results/listings.
     */
    @Throws(PluginFormatException::class)
    fun extractMetadata(input: InputStream): PluginMetadata
}
